dns/client.go

```diff
@@ -102,13 +148,141 @@ func (r *Client) Resolve(
 	return ips, nil
 }
 
-// Lookup resolves a domain name to a slice of IP addresses using the system's
-// default resolver.
+// Lookup resolves a domain name to a slice of IP addresses.
+//
+// It prefers the k6 VU dialer's resolver if available, so that k6 networking
+// options (e.g. blockHostnames) are honored. It falls back to the system's
+// default resolver otherwise.
 func (r *Client) Lookup(ctx context.Context, hostname string) ([]string, error) {
-	ips, err := net.DefaultResolver.LookupHost(ctx, hostname)
+	vustate := r.vu.State()
+	if vustate == nil {
+		return nil, common.NewInitContextError("using lookup in the init context is not supported")
+	}
+
+	// If k6's dialer is available, honor its blocked hostnames configuration.
+	if d, ok := vustate.Dialer.(*netext.Dialer); ok && d != nil && d.BlockedHostnames != nil {
+		normalized := strings.TrimSuffix(strings.ToLower(hostname), ".")
+		if _, blocked := d.BlockedHostnames.Contains(normalized); blocked {
+			return nil, &Error{
+				Name:    "BlockedHostname",
+				Message: fmt.Sprintf("blocked hostname: %s", normalized),
+				Kind:    Refused,
+			}
+		}
+	}
+
+	if vustate.Dialer == nil {
+		return nil, fmt.Errorf("unable to perform DNS lookup; reason: k6 internal network dialer is not available")
+	}
+	resolver := &net.Resolver{
+		Dial: vustate.Dialer.DialContext,
+	}
+
+	ips, err := resolver.LookupHost(ctx, hostname)
 	if err != nil {
-		return nil, fmt.Errorf("lookup of %s failed: %w", hostname, err)
+		return nil, &Error{
+			Name:    "LookupFailed",
+			Message: fmt.Sprintf("lookup of %s failed: %s", hostname, err.Error()),
+			Kind:    Refused,
+		}
 	}
 
 	return ips, nil
 }
+
+// k6DNSClient wraps dns.Client to use k6's dialer
+// This ensures k6's networking options (blockHostnames, blacklistIPs) are respected
+type k6DNSClient struct {
+	dns.Client
+	k6Dialer lib.DialContexter
+}
+
+// ExchangeContext overrides the default ExchangeContext to use k6's dialer
+func (c *k6DNSClient) ExchangeContext(
+	ctx context.Context,
+	m *dns.Msg,
+	address string,
+) (*dns.Msg, time.Duration, error) {
+	// If k6 dialer is not available, fall back to standard DNS client behavior
+	if c.k6Dialer == nil {
+		return c.Client.ExchangeContext(ctx, m, address)
+	}
+
+	start := time.Now()
+
+	// Create a connection using k6's dialer
+	conn, err := c.k6Dialer.DialContext(ctx, "udp", address)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer func() {
+		closeErr := conn.Close()
+		if closeErr != nil {
+			log.Fatalf("failed to close k6 DNS connection: %v", closeErr)
+		}
+	}()
+
+	// Set a reasonable deadline for the operation
+	var deadlineErr error
+	if deadline, ok := ctx.Deadline(); ok {
+		deadlineErr = conn.SetDeadline(deadline)
+	} else {
+		deadlineErr = conn.SetDeadline(time.Now().Add(5 * time.Second))
```

Comment:
Why an arbitrary value here?

Reply:
Ideally, I would like to get this value from k6, but to be absolutely transparent, I wasn't certain what the best way to do it would be. Considering we're doing UDP requests, 5 seconds already felt like an eternity.

Happy to adapt, and like mentioned into the other comment, I'm looking into the underlying dns library defaults 🙇🏻